betree/pmdk/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+                                     0, // Opening an existing file requires no flag(s).
+                                     0, // No length as no flag is provided.
+                                     0666, 
+                                     mapped_len, 
+                                     is_pmem);
+
+            match NonNull::new(ptr) {
+                //Some(value) => Some(ptr_to_pmem( Arc::new(Mutex::new(ptr)))),
+                Some(value) => Some(ptr_to_pmem( Arc::new(ptr))),
+                None => None
+            }
+        }
+    }
+
+    pub fn pmem_file_read(begin_pos: &ptr_to_pmem, offset: usize, data: &mut [u8], len: u64) -> Result<(), std::io::Error>{
+        //match NonNull::new(*begin_pos.0.lock().unwrap()) {
+        match NonNull::new(*begin_pos.0) {
+            None => {
+                return Err(std::io::Error::new(std::io::ErrorKind::Other,
+                                       format!("File handle is missing for the PMEM file.")
+                                       ));
+            },
+            _ => ()
+        };
+
+        unsafe {
+            //let ptr = pmem_memcpy(data.as_ptr() as *mut c_void, begin_pos.0.lock().unwrap().add(offset), len, PMEM_F_MEM_NOFLUSH /*| PMEM_F_MEM_TEMPORAL*/);
+            let ptr = pmem_memcpy(data.as_ptr() as *mut c_void, begin_pos.0.add(offset), len, PMEM_F_MEM_NOFLUSH /*| PMEM_F_MEM_TEMPORAL*/);
+            match NonNull::new(ptr) {
+                Some(value) => Ok(()),
+                None => Err(std::io::Error::new(std::io::ErrorKind::Other,
+                                                    format!("Failed to read data from  PMEM file. Offset: {}, Size:  {}",
+                                                            offset, len)))
+            };
+            Ok(())
+        }
+    }
+
+    pub fn pmem_file_write(begin_pos: &ptr_to_pmem, offset: usize, data: &[u8], len: usize) -> Result<(), std::io::Error>{
+        //match NonNull::new(*begin_pos.0.lock().unwrap()) {
+        match NonNull::new(*begin_pos.0) {
+            None => {
+                return Err(std::io::Error::new(std::io::ErrorKind::Other,
+                                       format!("File handle is missing for the PMEM file.")
+                                       ));
+            },
+            _ => ()
+        };
+
+        unsafe{
+            //let ptr = pmem_memcpy_persist( begin_pos.0.lock().unwrap().add(offset), data.as_ptr() as *mut c_void, len as u64);
+            let ptr = pmem_memcpy_persist( begin_pos.0.add(offset), data.as_ptr() as *mut c_void, len as u64);
+            match NonNull::new(ptr) {
+                Some(value) => Ok(()),
+                None => Err(std::io::Error::new(std::io::ErrorKind::Other,
+                                                    format!("Failed to write data to PMEM file. Offset: {}, Size:  {}",
+                                                            offset, len)))
+            };
+            Ok(())
+        }
+    }
```

Comment:
> Continuing from internal discussion.

To avoid locking overhead in this call we may want to consider leaving the raw pointer in the `Arc` as is but then we should probably mark the `pmem_file_write` function as `unsafe` to indicate that if two operations overlap the pmem user should be aware of race conditions that can occur here.